### PR TITLE
Randomly fuzz the PSM

### PR DESCRIPTION
**Pull request by tomaka**
My take on #2200 

Runs the peerset and randomly performs actions on it over time.

I'm aware of the fact that the tests fails, however I'm unsure whether it's the test that is wrong or if it's a bug in the PSM. If possible I'd be asking for reviews to see if what I'm doing is correct.

### CODE REVIEW

File: core/peerset/tests/fuzz.rs [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+			// Each of these weights corresponds to an action that we may perform.
+			let action_weights = [150, 90, 90, 30, 30, 1, 1, 4, 4];
+			match WeightedIndex::new(&action_weights).unwrap().sample(&mut rng) {
+				// If we generate 0, poll the peerset.
+				0 => match peerset.poll().unwrap() {
+					Async::Ready(Some(Message::Connect(id))) => {
+						if let Some(id) = incoming_nodes.iter().find(|(_, v)| **v == id).map(|(&id, _)| id) {
+							incoming_nodes.remove(&id);
+						}
+						assert!(connected_nodes.insert(id));
```
Comment (gterzian) [inline core/peerset/tests/fuzz.rs]: I think you might want to change this to `assert!(connected_nodes.insert(id) || known_nodes.remove(id));`, to account for an `Accept` message being received after `discovered` was called?

Comment (tomaka) [reply core/peerset/tests/fuzz.rs]: `known_nodes` contains all the nodes that have been discovered in the past, not just nodes waiting for connection.

**Comment by gavofyork** [comment]
@tomaka is this ready for merge?

**Comment by tomaka** [comment]
> @tomaka is this ready for merge?

No, the test failure is legitimate but I don't know if it's an issue in the tests or in the PSM.
I need to go back to this PR.

**Comment by tomaka** [comment]
Should be ready now.

**Comment by tomaka** [comment]
CI failure looks unrelated.

**Pull request merged**